File: ports/pyagentic/pyagentic/skills.py
"""Skills — a named bundle of (tools + a system-prompt fragment + required facts), the
portable analogue of the Flink ``Skill``. A path can declare ``skills: [billing]`` and
the builder expands them into the path's tool set + prompt, so capabilities compose
without rewiring each agent.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Sequence, Tuple


@dataclass(frozen=True)
class Skill:
    name: str
    tools: Tuple[str, ...] = ()
    prompt_fragment: str = ""
    required_facts: Tuple[str, ...] = ()


class SkillRegistry:
    """Named skills, looked up by the builder when a path lists ``skills``."""

    def __init__(self) -> None:
        self._skills: Dict[str, Skill] = {}

    def register(self, skill: Skill) -> "SkillRegistry":
        self._skills[skill.name] = skill
        return self

    def get(self, name: str) -> Skill:
        if name not in self._skills:
            raise KeyError(f"unknown skill {name!r}")
        return self._skills[name]

    def expand(self, names: Sequence[str]) -> Tuple[List[str], str, List[str]]:
        """Resolve skill names → (extra tool ids, joined prompt fragment, required facts)."""
        tools: List[str] = []
        fragments: List[str] = []
        facts: List[str] = []
        for n in names:
            s = self.get(n)
            for t in s.tools:
                if t not in tools:
                    tools.append(t)
            if s.prompt_fragment:
                fragments.append(s.prompt_fragment)
            facts.extend(s.required_facts)
        return tools, "\n".join(fragments), facts

    @classmethod
    def from_specs(cls, specs: List[dict]) -> "SkillRegistry":
        """Build from the YAML ``skills:`` list ([{name, tools, prompt, facts}])."""
        reg = cls()
        for s in specs or []:
            reg.register(Skill(
                name=s["name"],
                tools=tuple(s.get("tools", [])),
                prompt_fragment=s.get("prompt", ""),
                required_facts=tuple(s.get("facts", [])),
            ))
        return reg

File: ports/experimental/go/core/skills.go
package core

import "fmt"

// Skill is a named bundle of (tools + a system-prompt fragment + required facts) — the
// portable analogue of the Flink Skill. A path declares skills and the builder expands
// them into the path's tool set + prompt.
type Skill struct {
	Name           string
	Tools          []string
	PromptFragment string
	RequiredFacts  []string
}

// SkillRegistry holds named skills, looked up by the builder.
type SkillRegistry struct {
	skills map[string]Skill
}

// NewSkillRegistry builds an empty registry.
func NewSkillRegistry() *SkillRegistry { return &SkillRegistry{skills: map[string]Skill{}} }

// Register adds a skill and returns the registry for chaining.
func (r *SkillRegistry) Register(s Skill) *SkillRegistry {
	r.skills[s.Name] = s
	return r
}

// Get returns a skill (and whether it exists).
func (r *SkillRegistry) Get(name string) (Skill, bool) {
	s, ok := r.skills[name]
	return s, ok
}

// Expand resolves skill names → (extra tool ids, joined prompt fragment, required facts).
func (r *SkillRegistry) Expand(names []string) ([]string, string, []string) {
	var tools, facts []string
	var fragment string
	for _, n := range names {
		s, ok := r.skills[n]
		if !ok {
			continue
		}
		for _, t := range s.Tools {
			if !containsStr(tools, t) {
				tools = append(tools, t)
			}
		}
		if s.PromptFragment != "" {
			if fragment != "" {
				fragment += "\n"
			}
			fragment += s.PromptFragment
		}
		facts = append(facts, s.RequiredFacts...)
	}
	return tools, fragment, facts
}

// SkillRegistryFromSpecs builds from the YAML `skills:` list.
func SkillRegistryFromSpecs(specs []map[string]any) *SkillRegistry {
	reg := NewSkillRegistry()
	for _, s := range specs {
		name, _ := s["name"].(string)
		skill := Skill{Name: name, PromptFragment: fmt.Sprint(orEmpty(s["prompt"]))}
		for _, t := range toAnySlice(s["tools"]) {
			skill.Tools = append(skill.Tools, fmt.Sprint(t))
		}
		for _, f := range toAnySlice(s["facts"]) {
			skill.RequiredFacts = append(skill.RequiredFacts, fmt.Sprint(f))
		}
		reg.Register(skill)
	}
	return reg
}

func orEmpty(v any) any {
	if v == nil {
		return ""
	}
	return v
}

func toAnySlice(v any) []any {
	if l, ok := v.([]any); ok {
		return l
	}
	return nil
}

func containsStr(xs []string, x string) bool {
	for _, v := range xs {
		if v == x {
			return true
		}
	}
	return false
}
